src/main/java/org/jitsi/videobridge/LipSyncHack.java

```diff
@@ -0,0 +1,423 @@
+        -69, 46, 0, -2, -78, 38, 50, -16, 47, -126, -99, -32, 50, 32, 67, 100,
+        0};
+
+    /**
+     * A constant defining the maximum number of black key frames to send.
+     */
+    private static final int MAX_KEY_FRAMES = 20;
+
+    /**
+     * The rate (in ms) at which we are to send black key frames.
+     */
+    private static final long KEY_FRAME_RATE_MS = 33;
+
+    /**
+     * Wait for media for WAIT_MS before sending frames.
+     */
+    private final long WAIT_MS = 1000;
+
+    /**
+     * The <tt>Logger</tt> used by the <tt>LipSyncHack</tt> class and
+     * its instances to print debug information.
+     */
+    private static final Logger logger = Logger.getLogger(LipSyncHack.class);
+
+    /**
+     * The owner of this hack.
+     */
+    private final Endpoint endpoint;
+
+    /**
+     * The executor service that takes care of black key frame scheduling
+     * and injection.
+     */
+    private final ScheduledExecutorService scheduler =
+        Executors.newScheduledThreadPool(1);
+
+    /**
+     * The remote audio SSRCs that have been accepted by the translator and
+     * forwarded to the endpoint associated to this instance.
+     */
+    private final List<Long> acceptedAudioSSRCs = new ArrayList<>();
+
+    /**
+     * The remote video SSRCs that have been accepted by the translator and
+     * forwarded to the endpoint associated to this instance.
+     */
+    private final List<Long> acceptedVideoSSRCs = new ArrayList<>();
+
+    /**
+     * A map that holds all the inject states.
+     */
+    private final Map<Long, InjectState> states = new HashMap<>();
+
+    /**
+     * Ctor.
+     *
+     * @param endpoint
+     */
+    public LipSyncHack(Endpoint endpoint)
+    {
+        this.endpoint = endpoint;
+    }
+
+    /**
+     * Notifies this instance that an audio packet is about to be written.
+     *
+     * @param data
+     * @param buffer
+     * @param offset
+     * @param length
+     * @param source
+     */
+    public void onRTPTranslatorWillWriteAudio(
+        boolean data, byte[] buffer, int offset,
+        int length, Channel source)
+    {
+        if (!data)
+        {
+            return;
+        }
+
+        // Decide whether to trigger the hack or not.
+        Long acceptedAudioSSRC
+            = RawPacket.getSSRCAsLong(buffer, offset, length);
+
+        // In order to minimize the synchronization overhead, we process
+        // only the first data packet of a given RTP stream.
+        //
+        // XXX No synchronization is required to r/w the acceptedAudioSSRCs
+        // because in the current architecture this method is called by a single
+        // thread at the time.
+        if (acceptedAudioSSRCs.contains(acceptedAudioSSRC))
+        {
+            // We've already triggered the hack for this audio stream and its
+            // associated video streams.
+            return;
+        }
+
+        acceptedAudioSSRCs.add(acceptedAudioSSRC);
```

Comment:
Would it make sense to move this below the sanity checks (e.g. right before the for loop), in order to avoid any race condition with a video channel being created and added to its Endpoint? Or is that for the case where an endpoint only has an audio channel?